### PR TITLE
[Heng Raft]implement leader election

### PR DESCRIPTION
**Pull request by flyingmeat**

### CODE REVIEW

File: lab/src/raft/leader_election.go [hunk 1]
```diff
@@ -8,6 +8,21 @@ func (rf *Raft) SendHeartbeat() {
...
+		rf.closeVoteChan <- true
+	} else if rf.status == leader {
+		rf.clockInterval = time.Duration(rand.Intn(150) + 200 + (rf.me*100)%150)
+	}
+	rf.status = follower
+	rf.onEmptyElectionTerm = false
+	rf.SetHeatBeatClock()
+	rf.mu.Unlock()
+}
+
```
Comment (liana-yang) [inline lab/src/raft/leader_election.go:25]: Let's consider renaming `heartbeatClock` to `clock`.

File: lab/src/raft/leader_election.go [hunk 2]
```diff
@@ -31,11 +46,112 @@ func (rf *Raft) LeaderElection() {
...
+	for {
+		select {
+			case _, ok := <- rf.voteChan:
+				if !ok {
+					break
+				}
+				voteCount++
+				if voteCount > len(rf.log) / 2 {
+					// Change to leader
+					rf.mu.Lock()
```
Comment (flyingmeat) [inline lab/src/raft/leader_election.go:144]: move this logic to a function which make things clearer

File: lab/src/raft/leader_election.go (hunk 1 above)
Comment (flyingmeat) [reply lab/src/raft/leader_election.go:25]: Done

File: lab/src/raft/leader_election.go (hunk 2 above)
Comment (flyingmeat) [reply lab/src/raft/leader_election.go:144]: Done

Comment (flyingmeat) [reply lab/src/raft/leader_election.go:144]: There should be more comments to make code better understanding

**Review by liana-yang — APPROVED** [review]

**Pull request merged**